File: internal/plugins/deerkins/kins.go
package deerkins

import (
	"sort"
	"strings"
)

// The kinskode format is one character per cell, rows separated by \n. ' ' is
// black, 'A'-'O' the other fifteen mIRC colours, '_' transparent. It is the
// same format the web app stores, so the two render identically.
const (
	maxRows     = 30
	maxCols     = 40
	transparent = '_'

	fill      = "@"
	colorCode = "\x03"
	resetCode = "\x0f"
	boldCode  = "\x02"
)

var ircColor = map[byte]string{
	' ': "01", 'A': "00", 'B': "02", 'C': "03", 'D': "04", 'E': "05",
	'F': "06", 'G': "07", 'H': "08", 'I': "09", 'J': "10", 'K': "11",
	'L': "12", 'M': "13", 'N': "14", 'O': "15",
}

// inverse pairs each colour with its opposite for the invert modifier.
var inverse = map[byte]byte{
	' ': 'A', 'A': ' ', 'B': 'H', 'C': 'M', 'D': 'J', 'E': 'K',
	'F': 'I', 'G': 'L', 'H': 'B', 'I': 'M', 'J': 'D', 'K': 'E',
	'L': 'H', 'M': 'I', 'N': 'O', 'O': 'N', '_': ' ',
}

// modifiers are the single letters a request may carry before the pipe.
// 'x' is not a transform; it rolls a random pile of the others.
var modifierOrder = []byte{'i', 'm', 'n', 'd', 'r', 'u', 's', 'f', 't', 'x'}

var modifierNames = map[byte]string{
	'i': "invert", 'm': "mirror", 'n': "unitinu", 'd': "divide",
	'r': "reverse", 'u': "upsidedown", 's': "square", 'f': "flip",
	't': "transpose", 'x': "x",
}

var transforms = map[string]func([]string) []string{
	"invert":     invert,
	"mirror":     func(rows []string) []string { return halves(rows, 0) },
	"unitinu":    func(rows []string) []string { return halves(rows, 1) },
	"divide":     func(rows []string) []string { return halves(rows, 2) },
	"reverse":    reverse,
	"upsidedown": upsideDown,
	"square":     square,
	"flip":       flip,
	"transpose":  transposeRows,
}

// randomNames is the pool 'x' draws from, in a fixed order so a seeded
// generator gives the same picture twice.
var randomNames = []string{
	"invert", "mirror", "unitinu", "divide", "reverse",
	"upsidedown", "square", "flip", "transpose",
}

const (
	xRounds  = 3
	xMinMods = 4
	xMaxMods = 10
)

// normalise turns stored kinskode into rows of known characters. Anything the
// palette doesn't cover becomes transparent, so a row is always ascii and
// safe to index by byte. Oversized art is cropped rather than rejected: the
// gallery holds rows written long before the current limits.
func normalise(kinskode string) []string {
	kinskode = strings.ReplaceAll(kinskode, "\r\n", "\n")
	kinskode = strings.ReplaceAll(kinskode, "\r", "\n")

	rows := strings.Split(strings.ToUpper(kinskode), "\n")
	if len(rows) > maxRows {
		rows = rows[:maxRows]
	}

	out := make([]string, 0, len(rows))
	for _, row := range rows {
		var b strings.Builder
		for i := 0; i < len(row) && b.Len() < maxCols; i++ {
			if _, ok := ircColor[row[i]]; ok {
				b.WriteByte(row[i])
			} else {
				b.WriteByte(transparent)
			}
		}
		out = append(out, b.String())
	}
	for len(out) > 0 && isBlank(out[len(out)-1]) {
		out = out[:len(out)-1]
	}
	return out
}

func isBlank(row string) bool {
	return strings.Trim(row, string(transparent)) == ""
}

// toIRC paints each row with mIRC colour codes. Runs of one colour collapse to
// a bare fill and the colour state resets per line. A row stops early rather
// than exceed budget bytes, which keeps the line under the 512 byte protocol
// limit however wide the art is.
func toIRC(rows []string, budget int) []string {
	out := make([]string, 0, len(rows))
	for _, row := range rows {
		var b strings.Builder
		last := ""
		for i := 0; i < len(row); i++ {
			color, ok := ircColor[row[i]]
			if !ok {
				color = "00" // transparent paints white, as on the web
			}
			cell := fill
			if color != last {
				cell = colorCode + color + "," + color + fill
			}
			if b.Len()+len(cell)+len(resetCode) > budget {
				break
			}
			b.WriteString(cell)
			last = color
		}
		b.WriteString(resetCode)
		out = append(out, b.String())
	}
	return out
}

// clamp bounds the art after modifiers have run. flip turns columns into rows,
// so a wide drawing can come back taller than it went in.
func clamp(rows []string, maxLines int) []string {
	if maxLines > maxRows {
		maxLines = maxRows
	}
	if len(rows) > maxLines {
		rows = rows[:maxLines]
	}
	out := make([]string, len(rows))
	for i, row := range rows {
		out[i] = head(row, maxCols)
	}
	return out
}

// splitRequest cuts "mods|name" into its parts. Modifiers are deduplicated and
// sorted, and 'x' swallows everything else.
func splitRequest(request string) (mods []byte, name string) {
	if i := strings.IndexByte(request, '|'); i >= 0 {
		mods, name = parseMods(request[:i]), request[i+1:]
	} else {
		name = request
	}
	return mods, strings.TrimSpace(name)
}

func parseMods(raw string) []byte {
	seen := map[byte]bool{}
	var mods []byte
	for i := 0; i < len(raw); i++ {
		c := lowerByte(raw[i])
		if _, ok := modifierNames[c]; !ok || seen[c] {
			continue
		}
		if c == 'x' {
			return []byte{'x'}
		}
		seen[c] = true
		mods = append(mods, c)
	}
	sort.Slice(mods, func(i, j int) bool { return mods[i] < mods[j] })
	return mods
}

func lowerByte(c byte) byte {
	if c >= 'A' && c <= 'Z' {
		return c + ('a' - 'A')
	}
	return c
}

// applyMods runs the requested modifiers and reports the names of the ones
// that actually ran, which is the only way to know what 'x' rolled.
func applyMods(rows []string, mods []byte, roll func(int) int) ([]string, []string) {
	if len(mods) == 1 && mods[0] == 'x' {
		return applyX(rows, roll)
	}
	var used []string
	for _, m := range mods {
		fn, ok := transforms[modifierNames[m]]
		if !ok {
			continue
		}
		rows = fn(rows)
		used = append(used, modifierNames[m])
	}
	return rows, used
}

// applyX rolls between xMinMods and xMaxMods modifiers, then flips a coin on
// whether to go round again, up to xRounds times.
func applyX(rows []string, roll func(int) int) ([]string, []string) {
	var used []string
	for round := 0; round < xRounds; round++ {
		n := xMinMods + roll(xMaxMods-xMinMods+1)
		for i := 0; i < n; i++ {
			name := randomNames[roll(len(randomNames))]
			rows = transforms[name](rows)
			used = append(used, name)
		}
		if roll(2) == 0 {
			break
		}
	}
	return rows, used
}

func invert(rows []string) []string {
	out := make([]string, len(rows))
	for i, row := range rows {
		b := []byte(row)
		for j := range b {
			if c, ok := inverse[b[j]]; ok {
				b[j] = c
			}
		}
		out[i] = string(b)
	}
	return out
}

func reverse(rows []string) []string {
	out := make([]string, len(rows))
	for i, row := range rows {
		out[i] = reverseString(row)
	}
	return out
}

func upsideDown(rows []string) []string {
	out := make([]string, len(rows))
	for i, row := range rows {
		out[len(rows)-1-i] = row
	}
	return out
}

// halves mirrors one half of every row onto the other. direction 0 folds the
// left half rightwards, 1 the right half leftwards, 2 swaps the two.
func halves(rows []string, direction int) []string {
	half := widest(rows) / 2
	if half < 1 {
		return rows
	}
	out := make([]string, len(rows))
	for i, row := range rows {
		rev := reverseString(row)
		switch direction {
		case 0:
			out[i] = head(rev, half) + tail(row, half)
		case 1:
			out[i] = head(row, half) + tail(rev, half)
		default:
			out[i] = tail(rev, half) + head(row, half)
		}
	}
	return out
}

// square rotates every row about its centre and the rows about theirs.
func square(rows []string) []string {
	half := widest(rows) / 2
	if half < 1 {
		return rows
	}
	out := make([]string, 0, len(rows))
	for i, row := range rows {
		shifted := tail(row, half) + head(row, half)
		if i < len(rows)/2 {
			out = append(out, shifted)
		} else {
			out = append([]string{shifted}, out...)
		}
	}
	return out
}

// flip turns columns into rows.
func flip(rows []string) []string {
	width := widest(rows)
	if width == 0 {
		return rows
	}
	out := make([]string, 0, width)
	for x := 0; x < width; x++ {
		b := make([]byte, len(rows))
		for y, row := range rows {
			if x < len(row) {
				b[y] = row[x]
			} else {
				b[y] = transparent
			}
		}
		out = append(out, string(b))
	}
	return out
}

// transposeRows shears the picture, rotating each row one cell further than
// the one above it. The shift counter restarts whenever it reaches the width of
// the row it is about to move, which is why ragged art shears unevenly. The web
// app never ported this one, so the original is the only definition of it.
func transposeRows(rows []string) []string {
	out := make([]string, len(rows))
	shift := 0
	for i, row := range rows {
		width := len(row)
		if shift == width {
			shift = 0
		}
		if shift == 0 || width == 0 {
			out[i] = row
			shift++
			continue
		}
		keep := width - shift
		if keep < 0 {
			keep = width + keep // a shift past the width eats into the end
		}
		out[i] = tail(row, width-shift) + head(row, keep)
		shift++
	}
	return out
}

func widest(rows []string) int {
	w := 0
	for _, row := range rows {
		if len(row) > w {
			w = len(row)
		}
	}
	return w
}

func reverseString(s string) string {
	b := []byte(s)
	for i, j := 0, len(b)-1; i < j; i, j = i+1, j-1 {
		b[i], b[j] = b[j], b[i]
	}
	return string(b)
}

// head and tail slice without panicking on an index past the end, the way the
// web app's slice() does.
func head(s string, n int) string {
	if n >= len(s) {
		return s
	}
	if n < 0 {
		return ""
	}
	return s[:n]
}

func tail(s string, n int) string {
	if n >= len(s) {
		return ""
	}
	if n < 0 {
		return s
	}
	return s[n:]
}

File: web/public/deerkins/kins.js
/*
 * The kinskode format: one character per cell, lines separated by \n.
 *   ' '      black (IRC colour 01)
 *   'A'-'O'  the other 15 mIRC colours
 *   '_'      transparent
 *
 */

export const MAX_ROWS = 30;
export const MAX_COLS = 40;

export const TRANSPARENT = "_";

export const PALETTE = [
	{ kins: " ", irc: "01", hex: "#000000", name: "black" },
	{ kins: "A", irc: "00", hex: "#FFFFFF", name: "white" },
	{ kins: "B", irc: "02", hex: "#00007F", name: "navy" },
	{ kins: "C", irc: "03", hex: "#009300", name: "green" },
	{ kins: "D", irc: "04", hex: "#FF0000", name: "red" },
	{ kins: "E", irc: "05", hex: "#7F0000", name: "maroon" },
	{ kins: "F", irc: "06", hex: "#9C009C", name: "purple" },
	{ kins: "G", irc: "07", hex: "#FC7E00", name: "orange" },
	{ kins: "H", irc: "08", hex: "#FFFF00", name: "yellow" },
	{ kins: "I", irc: "09", hex: "#00FC00", name: "lime" },
	{ kins: "J", irc: "10", hex: "#009393", name: "teal" },
	{ kins: "K", irc: "11", hex: "#00FFFF", name: "cyan" },
	{ kins: "L", irc: "12", hex: "#0000FC", name: "blue" },
	{ kins: "M", irc: "13", hex: "#FF00FF", name: "magenta" },
	{ kins: "N", irc: "14", hex: "#7F7F7F", name: "grey" },
	{ kins: "O", irc: "15", hex: "#D2D2D2", name: "silver" },
];

export const KINS_CHARS = PALETTE.map((c) => c.kins).concat(TRANSPARENT);

const HEX_BY_KINS = new Map(PALETTE.map((c) => [c.kins, c.hex]));
const IRC_BY_KINS = new Map(PALETTE.map((c) => [c.kins, c.irc]));
const KINS_BY_IRC = new Map(PALETTE.map((c) => [c.irc, c.kins]));

export function hexOf(char) {
	return HEX_BY_KINS.get(char) ?? null;
}

const COLOR = "\x03"; // mIRC colour prefix (^C)
const RESET = "\x0f"; // mIRC reset (^O)
const FILL = "@";

export function normalise(kinskode) {
	const lines = String(kinskode)
		.replace(/\r\n?/g, "\n")
		.toUpperCase()
		.split("\n")
		.map((line) =>
			Array.from(line, (ch) =>
				KINS_CHARS.includes(ch) ? ch : TRANSPARENT,
			).join(""),
		);

	while (lines.length && /^_*$/.test(lines[lines.length - 1])) lines.pop();
	return lines.join("\n");
}

/** Returns null when safe to store, otherwise the reason. Run on the server. */
export function validate(kinskode) {
	if (typeof kinskode !== "string") return "kinskode must be a string";
	if (kinskode.length === 0) return "kinskode is empty";
	if (kinskode.length > (MAX_COLS + 1) * MAX_ROWS)
		return "kinskode is too large";

	const lines = kinskode.split("\n");
	if (lines.length > MAX_ROWS) return `too many rows (max ${MAX_ROWS})`;

	for (const line of lines) {
		if (line.length > MAX_COLS) return `line too long (max ${MAX_COLS})`;
		for (const ch of line) {
			if (!KINS_CHARS.includes(ch))
				return "kinskode contains an illegal character";
		}
	}

	if (!/[^_\n]/.test(kinskode)) return "the canvas is blank";
	return null;
}

/** Repeated cells collapse to a bare '@'. Colour state resets per line. */
export function toIrc(kinskode) {
	return kinskode
		.split("\n")
		.map((line) => {
			let out = "";
			let last = null;
			for (const ch of line) {
				const irc = IRC_BY_KINS.get(ch) ?? "00"; // transparent renders as white
				if (irc === last) {
					out += FILL;
				} else {
					out += `${COLOR}${irc},${irc}${FILL}`;
					last = irc;
				}
			}
			return out + RESET;
		})
		.join("\n");
}

export function fromIrc(irccode) {
	return String(irccode)
		.replace(/\r\n?/g, "\n")
		.split("\n")
		.map((line) => {
			let out = "";
			let current = " ";
			// ^Cnn[,nn] sets the colour; nn,nn without ^C does too; @ paints a cell.
			const token = /\x03(\d{1,2})(?:,(\d{1,2}))?|(\d{1,2}),(\d{1,2})|@|\x0f/g;
			let m;
			// biome-ignore lint/suspicious/noAssignInExpressions: the standard exec loop
			while ((m = token.exec(line)) !== null) {
				if (m[0] === FILL) {
					out += out.length < MAX_COLS ? current : "";
				} else if (m[0] === RESET) {
					current = " ";
				} else {
					const bg = m[2] ?? m[4] ?? m[1] ?? m[3];
					current = KINS_BY_IRC.get(String(bg).padStart(2, "0")) ?? " ";
				}
			}
			return out;
		})
		.join("\n");
}

export function autoCrop(kinskode) {
	const lines = kinskode.split("\n");
	let top = 0;
	let bottom = lines.length - 1;
	while (top <= bottom && /^_*$/.test(lines[top])) top++;
	while (bottom >= top && /^_*$/.test(lines[bottom])) bottom--;
	if (top > bottom) return "";

	const rows = lines.slice(top, bottom + 1);
	const width = Math.max(...rows.map((r) => r.length));
	let left = 0;
	let right = width - 1;
	const columnBlank = (x) =>
		rows.every((r) => (r[x] ?? TRANSPARENT) === TRANSPARENT);
	while (left <= right && columnBlank(left)) left++;
	while (right >= left && columnBlank(right)) right--;

	return rows
		.map((r) => r.padEnd(width, TRANSPARENT).slice(left, right + 1))
		.join("\n");
}

export function measure(kinskode) {
	const lines = kinskode.split("\n");
	return {
		rows: lines.length,
		cols: Math.max(0, ...lines.map((l) => l.length)),
	};
}

const INVERSE = {
	" ": "A",
	A: " ",
	B: "H",
	C: "M",
	D: "J",
	E: "K",
	F: "I",
	G: "L",
	H: "B",
	I: "M",
	J: "D",
	K: "E",
	L: "H",
	M: "I",
	N: "O",
	O: "N",
	_: " ",
};

const split = (k) => k.split("\n");
const join = (l) => l.join("\n");
const widest = (lines) => Math.max(0, ...lines.map((l) => l.length));

export const transforms = {
	/** i: swap every colour for its opposite. */
	invert: (k) =>
		Array.from(k, (ch) => (ch === "\n" ? ch : (INVERSE[ch] ?? ch))).join(""),

	/** r: flip horizontally. */
	reverse: (k) => join(split(k).map((l) => Array.from(l).reverse().join(""))),

	/** u: flip vertically. */
	upsidedown: (k) => join(split(k).reverse()),

	/** f: transpose (rows become columns). */
	flip: (k) => {
		const lines = split(k);
		const width = widest(lines);
		const out = [];
		for (let x = 0; x < width; x++) {
			out.push(lines.map((l) => l[x] ?? TRANSPARENT).join(""));
		}
		return out.length ? join(out) : k;
	},

	/** m: mirror the left half onto the right. */
	mirror: (k) => halves(k, 0),

	/** n: mirror the right half onto the left. */
	unitinu: (k) => halves(k, 1),

	/** d: swap the halves, reversing the left one. */
	divide: (k) => halves(k, 2),

	/** s: rotate each row about its centre, and the rows about theirs. */
	square: (k) => {
		const lines = split(k);
		const half = Math.floor(widest(lines) / 2);
		if (half < 1) return k;
		const out = [];
		lines.forEach((line, i) => {
			const shifted = line.slice(half) + line.slice(0, half);
			if (i < Math.floor(lines.length / 2)) out.push(shifted);
			else out.unshift(shifted);
		});
		return join(out);
	},
};

function halves(kinskode, direction) {
	const lines = split(kinskode);
	const half = Math.floor(widest(lines) / 2);
	if (half < 1) return kinskode;

	const rev = (s) => Array.from(s).reverse().join("");
	return join(
		lines.map((line) => {
			if (direction === 0) return rev(line).slice(0, half) + line.slice(half);
			if (direction === 1) return line.slice(0, half) + rev(line).slice(half);
			return rev(line).slice(half) + line.slice(0, half);
		}),
	);
}

export const MAX_NAME = 48;
export const MAX_CREATOR = 32;

export function sanitiseName(name) {
	return String(name ?? "")
		.toLowerCase()
		.replace(/[^a-z0-9 \-_]/g, "")
		.replace(/\s+/g, " ")
		.trim()
		.slice(0, MAX_NAME);
}

export function sanitiseCreator(creator) {
	return String(creator ?? "")
		.replace(/[\p{C}<>]/gu, "")
		.replace(/\s+/g, " ")
		.trim()
		.slice(0, MAX_CREATOR);
}

export function nameSuffix() {
	const bytes = crypto.getRandomValues(new Uint8Array(5));
	return Array.from(bytes, (b) => String.fromCharCode(97 + (b % 26))).join("");
}
